### PR TITLE
Verify sha256/sha1 of downloaded files (#3014)

### PR DESCRIPTION
**Pull request by sbbroot**

### CODE REVIEW

**Comment by przemyslavic** [comment]
I believe this should be fixed first: https://github.com/epiphany-platform/epiphany/issues/3028.

And it doesn't work properly. It shows there is an error but still does not terminate the script and the file is downloaded despite the message that it could not be downloaded.

```
2022-03-23 10:15:16,017 [INFO]: Downloading images...
2022-03-23 10:15:16,190 [DEBUG]: - haproxy:2.2.2-alpine - checksum ok, skipped
2022-03-23 10:15:17,852 [DEBUG]: - kubernetesui/dashboard:v2.3.1 - checksum ok, skipped
2022-03-23 10:15:18,120 [DEBUG]: - kubernetesui/metrics-scraper:v1.0.7 - checksum ok, skipped
2022-03-23 10:15:18,210 [INFO]: - registry:2
2022-03-23 10:15:18,210 [DEBUG]: [1/3] Running: `crane pull --insecure --platform=linux/amd64 --format=legacy registry:2 /tmp/tmp1onii_en`
2022-03-23 10:15:23,850 [ERROR]: - registry:2 - download failed due to checksum mismatch, WARNING someone might have replaced the file
2022-03-23 10:15:23,851 [WARNING]: Could not download image: `registry:2`
2022-03-23 10:15:25,139 [DEBUG]: - bitnami/pgpool:4.2.4 - checksum ok, skipped
2022-03-23 10:15:26,399 [DEBUG]: - bitnami/pgbouncer:1.16.0 - checksum ok, skipped
2022-03-23 10:15:32,080 [DEBUG]: - epiphanyplatform/keycloak:14.0.0 - checksum ok, skipped
2022-03-23 10:15:33,405 [DEBUG]: - rabbitmq:3.8.9 - checksum ok, skipped
```

**Review by erzetpe — DISMISSED** [review]

**Review by seriva — DISMISSED** [review]

**Review by seriva — DISMISSED** [review]

**Comment by przemyslavic** [comment]
1. :x: This breaks the upgrade:
```
2022-04-01T08:42:13.9263227Z[38;21m08:42:13 INFO cli.src.ansible.AnsibleCommand - TASK [image_registry : Run registry] *******************************************
2022-04-01T08:42:14.6127202Z[31;21m08:42:14 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-10todevazrhflannel-repository-vm-0]: FAILED! => {"changed": true, "cmd": ["docker", "run", "-d", "-e", "REGISTRY_HTTP_ADDR=0.0.0.0:5000", "-p", "5000:5000", "--restart=always", "--name", "epiphany-registry", "registry:2.8.0"], "delta": "0:00:00.055780", "end": "2022-04-01 08:42:14.531076", "msg": "non-zero return code", "rc": 125, "start": "2022-04-01 08:42:14.475296", "stderr": "docker: Error response from daemon: Conflict. The container name \"/epiphany-registry\" is already in use by container \"551d80aee089ca7389ab50b51b5627262bd4e6fd656689d83796e2165b446249\". You have to remove (or rename) that container to be able to reuse that name.\nSee 'docker run --help'.", "stderr_lines": ["docker: Error response from daemon: Conflict. The container name \"/epiphany-registry\" is already in use by container \"551d80aee089ca7389ab50b51b5627262bd4e6fd656689d83796e2165b446249\". You have to remove (or rename) that container to be able to reuse that name.", "See 'docker run --help'."], "stdout": "", "stdout_lines": []}
```

2. {image} variable is not resolved
```
2022-04-01 10:37:26,878 [WARNING]: - {image} - allow_mismatch flag used, continue downloading
```

**Comment by przemyslavic** [comment]
1. there are no `-q/-f` flags in `docker kill` command
```
2022-04-07T13:07:20.2360320Z[31;21m13:07:20 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-13todevazrhflannel-repository-vm-0]: FAILED! => {"changed": true, "cmd": ["docker", "kill", "-q", "-f", "registry:2"], "delta": "0:00:00.089284", "end": "2022-04-07 13:07:20.150192", "msg": "non-zero return code", "rc": 125, "start": "2022-04-07 13:07:20.060908", "stderr": "unknown shorthand flag: 'q' in -q\nSee 'docker kill --help'.", "stderr_lines": ["unknown shorthand flag: 'q' in -q", "See 'docker kill --help'."], "stdout": "", "stdout_lines": []}
```
2. still to be fixed
3. typo in `current_regitry_up_check`
4. jinja2 templating warning:
```
2022-04-07T13:07:19.5135338Z[38;21m13:07:19 INFO cli.src.ansible.AnsibleCommand - TASK [image_registry : Kill previous version running registry] *****************
2022-04-07T13:07:19.5317946Z[33;21m13:07:19 WARNING cli.src.ansible.AnsibleCommand - [WARNING]: conditional statements should not include jinja2 templating
2022-04-07T13:07:19.5319794Z[38;21m13:07:19 INFO cli.src.ansible.AnsibleCommand - delimiters such as {{ }} or {% %}. Found: specification.registry_image.name !=
2022-04-07T13:07:19.5320822Z[38;21m13:07:19 INFO cli.src.ansible.AnsibleCommand - '{{ running_registry_images.stdout }}'
```
5. shouldn't we try `docker stop` instead of `docker kill` first?

File: ansible/playbooks/roles/image_registry/tasks/main.yml [hunk 1]
```diff
@@ -36,16 +36,18 @@
...
       block:
         - name: Get current running registry container
           command: "{% raw %}docker ps --format '{{ .Image }}' -f 'name=registry' -f 'status=running'{% endraw %}"
           register: running_registry_images
 
-        - name: Kill previous version running registry
-          command: docker kill -q -f '{{ running_registry_images.stdout }}'
-          when: specification.registry_image.name != '{{ running_registry_images.stdout }}'
+        - name: Stop previous version running registry
+          command: docker stop '{{ running_registry_images.stdout }}'
```
Comment (przemyslavic) [inline ansible/playbooks/roles/image_registry/tasks/main.yml]: It won't work anyway. `running_registry_images.stdout` contains the name of the image, and the `docker stop` command should be executed against the container, not the image.
You should also probably register the output from `docker ps --format '{{ .Names }}' -f 'name=registry' -f 'status=running'` and then use it in the `docker stop` command.  Or use `epiphany-registry` name which is hardcoded: https://github.com/epiphany-platform/epiphany/pull/3036/files#diff-863026b2f688ec8ed844ce4adb991e74ab6fd30095c809b6a444d6aac562fa67R69
Or some other clever way. :)

**Comment by przemyslavic** [comment]
:x: It still does not work properly
```
2022-04-11T12:39:40.6227280Z[38;21m12:39:40 INFO cli.src.ansible.AnsibleCommand - TASK [image_registry : Run registry] *******************************************[0m
2022-04-11T12:39:41.3321588Z[31;21m12:39:41 ERROR cli.src.ansible.AnsibleCommand - fatal: [ci-13todevazrhflannel-repository-vm-0]: FAILED! => {"changed": true, "cmd": ["docker", "run", "-d", "-e", "REGISTRY_HTTP_ADDR=0.0.0.0:5000", "-p", "5000:5000", "--restart=always", "--name", "epiphany-registry", "registry:2.8.0"], "delta": "0:00:00.045180", "end": "2022-04-11 12:39:41.247016", "msg": "non-zero return code", "rc": 125, "start": "2022-04-11 12:39:41.201836", "stderr": "docker: Error response from daemon: Conflict. The container name \"/epiphany-registry\" is already in use by container \"c2c0e6b4199cc3843a31d4e058fb02e7412aa48b152ade918381c3cb4008b920\". You have to remove (or rename) that container to be able to reuse that name.\nSee 'docker run --help'.", "stderr_lines": ["docker: Error response from daemon: Conflict. The container name \"/epiphany-registry\" is already in use by container \"c2c0e6b4199cc3843a31d4e058fb02e7412aa48b152ade918381c3cb4008b920\". You have to remove (or rename) that container to be able to reuse that name.", "See 'docker run --help'."], "stdout": "", "stdout_lines": []}
```
There should be `docker rm` command executed after `docker stop` or `docker rm -f` instead of these two.

**Review by przemyslavic — APPROVED** [review]

**Review by erzetpe — APPROVED** [review]

**Pull request merged**